cpp/include/raft/cluster/detail/kmeans_balanced.cuh

```diff
@@ -0,0 +1,1084 @@
+  auto [max_minibatch_size, mem_per_row] =
+    calc_minibatch_size<MathT>(n_clusters, n_rows, dim, params.metric, std::is_same_v<T, MathT>);
+  auto pool_guard =
+    raft::get_pool_memory_resource(device_memory, mem_per_row * size_t(max_minibatch_size));
+  if (pool_guard) {
+    RAFT_LOG_DEBUG("build_hierarchical: using pool memory resource with initial size %zu bytes",
+                   pool_guard->pool_size());
+  }
+
+  // Precompute the L2 norm of the dataset if relevant.
+  const MathT* dataset_norm = nullptr;
+  rmm::device_uvector<MathT> dataset_norm_buf(0, stream, device_memory);
+  if (params.metric == raft::distance::DistanceType::L2Expanded ||
+      params.metric == raft::distance::DistanceType::L2SqrtExpanded) {
+    dataset_norm_buf.resize(n_rows, stream);
+    for (IdxT offset = 0; offset < n_rows; offset += max_minibatch_size) {
+      IdxT minibatch_size = std::min<IdxT>(max_minibatch_size, n_rows - offset);
+      compute_norm(handle,
+                   dataset_norm_buf.data() + offset,
+                   dataset + dim * offset,
+                   dim,
+                   minibatch_size,
+                   mapping_op,
+                   device_memory);
+    }
+    dataset_norm = (const MathT*)dataset_norm_buf.data();
+  }
+
+  /* Temporary workaround to cub::DeviceHistogram not supporting any type that isn't natively
+   * supported by atomicAdd: find a supported CounterT based on the IdxT. */
+  typedef typename std::conditional_t<sizeof(IdxT) == 8, unsigned long long int, unsigned int>
+    CounterT;
+
+  // build coarse clusters (mesoclusters)
+  rmm::device_uvector<LabelT> mesocluster_labels_buf(n_rows, stream, &managed_memory);
+  rmm::device_uvector<CounterT> mesocluster_sizes_buf(n_mesoclusters, stream, &managed_memory);
+  {
+    rmm::device_uvector<MathT> mesocluster_centers_buf(n_mesoclusters * dim, stream, device_memory);
+    build_clusters(handle,
+                   params,
+                   dim,
+                   dataset,
+                   n_rows,
+                   n_mesoclusters,
+                   mesocluster_centers_buf.data(),
+                   mesocluster_labels_buf.data(),
+                   mesocluster_sizes_buf.data(),
+                   mapping_op,
+                   device_memory,
+                   dataset_norm);
+  }
+
+  auto mesocluster_sizes  = mesocluster_sizes_buf.data();
+  auto mesocluster_labels = mesocluster_labels_buf.data();
+
+  handle.sync_stream(stream);
+
+  // build fine clusters
+  auto [mesocluster_size_max, fine_clusters_nums_max, fine_clusters_nums, fine_clusters_csum] =
+    arrange_fine_clusters(n_clusters, n_mesoclusters, n_rows, mesocluster_sizes);
+
+  if (mesocluster_size_max * n_mesoclusters > 2 * n_rows) {
+    RAFT_LOG_WARN("build_hierarchical: built unbalanced mesoclusters");
+    RAFT_LOG_TRACE_VEC(mesocluster_sizes, n_mesoclusters);
+    RAFT_LOG_TRACE_VEC(fine_clusters_nums.data(), n_mesoclusters);
+  }
```

Comment:
Please merge the changes from https://github.com/rapidsai/raft/pull/1161, that's an important workaround to avoid out-of-memory when the clusters are very unbalanced.

Reply:
I just fixed that, perhaps double-check that I didn't forget anything? Thanks

Reply:
Yep, looks ok.